kedro-datasets/kedro_datasets/api/api_dataset.py

```diff
@@ -233,3 +234,21 @@ def _exists(self) -> bool:
         with sessions.Session() as session:
             response = self._execute_request(session)
         return response.ok
+
+
+_DEPRECATED_CLASSES = {
+    "APIDataSet": APIDataset,
+}
+
+
+def __getattr__(name):
+    if name in _DEPRECATED_CLASSES:
+        alias = _DEPRECATED_CLASSES[name]
+        warnings.warn(
```

Comment:
When is the warning supposed to be shown? I added similar changed to the `CSVDataSet` to test locally, but I don't get the message when using the "old" spelling in the catalog or directly in a node.

Reply:
@merelcht Have you updated your warnings filter? `DeprecationWarning` is ignored by default; depending on how you're testing, you'd need to enable `default::DeprecationWarning`.

Reply:
Bear in mind that `DeprecationWarning` are only shown in `__main__` by default:

> In Python 2.7 and Python 3.2, the default warning filters were updated to hide DeprecationWarning by default [...]
However, this change has had the unfortunate side effect of making DeprecationWarning markedly less effective at its primary intended purpose: providing advance notice of breaking changes in APIs (whether in CPython, the standard library, or in third party libraries) to users of those APIs.

https://peps.python.org/pep-0565/

(Taken from https://github.com/kedro-org/kedro/issues/2744)

In fact, I noticed that parts of Kedro are using `FutureWarning` now for deprecated behavior, see https://github.com/kedro-org/kedro/commit/b3637fb49025292aa5afdf554a0090c3dc0a2bd9 and https://github.com/kedro-org/kedro/commit/3a02aa44b31b24fa75c9be3cc3ccfdc428cffdb1

Reply:
For example:

```python
(kedro-datasets) deepyaman@Deepyamans-MacBook-Air kedro-datasets % python -W default::DeprecationWarning
Python 3.10.9 (main, Mar  8 2023, 04:44:30) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from kedro_datasets.api import APIDataset
/opt/miniconda3/envs/kedro-datasets/lib/python3.10/site-packages/kedro/io/__init__.py:6: DeprecationWarning: 'CachedDataSet' has been renamed to 'CachedDataset', and the alias will be removed in Kedro 0.19.0
  from .cached_dataset import CachedDataset, CachedDataSet
/opt/miniconda3/envs/kedro-datasets/lib/python3.10/site-packages/kedro/io/__init__.py:16: DeprecationWarning: 'LambdaDataSet' has been renamed to 'LambdaDataset', and the alias will be removed in Kedro 0.19.0
  from .lambda_dataset import LambdaDataset, LambdaDataSet
/opt/miniconda3/envs/kedro-datasets/lib/python3.10/site-packages/kedro/io/__init__.py:17: DeprecationWarning: 'MemoryDataSet' has been renamed to 'MemoryDataset', and the alias will be removed in Kedro 0.19.0
  from .memory_dataset import MemoryDataset, MemoryDataSet
/opt/miniconda3/envs/kedro-datasets/lib/python3.10/site-packages/kedro/io/__init__.py:18: DeprecationWarning: 'IncrementalDataSet' has been renamed to 'IncrementalDataset', and the alias will be removed in Kedro 0.19.0
  from .partitioned_dataset import (
/opt/miniconda3/envs/kedro-datasets/lib/python3.10/site-packages/kedro/io/__init__.py:18: DeprecationWarning: 'PartitionedDataSet' has been renamed to 'PartitionedDataset', and the alias will be removed in Kedro 0.19.0
  from .partitioned_dataset import (
>>> from kedro_datasets.api import APIDataSet
/opt/miniconda3/envs/kedro-datasets/lib/python3.10/site-packages/lazy_loader/__init__.py:78: DeprecationWarning: 'APIDataSet' has been renamed to 'APIDataset', and the alias will be removed in Kedro-Datasets 2.0.0
  attr = getattr(submod, name)
>>> 
(kedro-datasets) deepyaman@Deepyamans-MacBook-Air kedro-datasets % python                               
Python 3.10.9 (main, Mar  8 2023, 04:44:30) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from kedro_datasets.api import APIDataset
>>> from kedro_datasets.api import APIDataSet
>>> 
```

Now, there is the separate issue of showing all of the deprecation warnings coming from core, if you turn on the warnings filter. I believe @astrojuanlu has pointed that out before. But, if we start with the goal of this being to raise warnings in the right place, I think we're good.

On the other issue, I can try to take a look later, but maybe using `lazy_loader` can help make sure all the other imports don't get triggered when not used.

Reply:
> Bear in mind that `DeprecationWarning` are only shown in `__main__` by default:
> 
> > In Python 2.7 and Python 3.2, the default warning filters were updated to hide DeprecationWarning by default [...]
> > However, this change has had the unfortunate side effect of making DeprecationWarning markedly less effective at its primary intended purpose: providing advance notice of breaking changes in APIs (whether in CPython, the standard library, or in third party libraries) to users of those APIs.
> 
> https://peps.python.org/pep-0565/
> 
> (Taken from [kedro-org/kedro#2744](https://github.com/kedro-org/kedro/issues/2744))
> 
> In fact, I noticed that parts of Kedro are using `FutureWarning` now for deprecated behavior, see [kedro-org/kedro@b3637fb](https://github.com/kedro-org/kedro/commit/b3637fb49025292aa5afdf554a0090c3dc0a2bd9) and [kedro-org/kedro@3a02aa4](https://github.com/kedro-org/kedro/commit/3a02aa44b31b24fa75c9be3cc3ccfdc428cffdb1)

@astrojuanlu TIL. Should we use `FutureWarning` here instead?

However--more a side note than anything--this would make it even more imperative that address the issue of making the warnings only occur on use for `PartitionedDataSet`, etc.

Reply:
@astrojuanlu Reading into it a bit more, I don't think `FutureWarning` sounds right, in line with their descriptions:

* `DeprecationWarning`: Base category for warnings about deprecated features when those warnings are intended for other Python developers.
* `FutureWarning`: Base category for warnings about deprecated features when those warnings are intended for end users of applications that are written in Python.

Our warnings really are intended for other Python developers (usually data scientists or engineers), not for end users of applications written in Python (e.g. some companies build end-to-end offerings and use Kedro for the data processing, in which case the end users--likely unaware that Kedro is even being used--shouldn't be alerted about what will be removed in a future version). Similarly, in past QB/McK projects, I've had EMs or clients who are shown how to run the pipeline (but not expected to modify anything); they, too, don't need to see the `DeprecationWarnings`s/it's not for them. In this latter case, they are interacting with it through the CLI, but they're not a Python developer.

I think this requires further discussion, and is perhaps outside the scope of this particular PR (since the chosen strategy should be applied universally, including in Kedro core). As such, I'm going to leave it as `DeprecationWarning`s for now, and we can do a bulk find/replace later, if necessary.

Reply:
Technically, you're right @deepyaman and the `DeprecationWarning` is what we should use here. However, I think in practice this means that only people who do explicit warning control will see these messages. Because datasets are such a big part of Kedro and are used in every Kedro project, I think it's very important for all our users to see these messages and give them time to migrate the dataset naming before their pipelines start failing. So I agree with @astrojuanlu that we should use `FutureWarning` for this specific use case. 

Ideally, we'll release `kedro-dataset` `2.0.0` very soon after this so if users find the warning really annoying they can just rename all their datasets and update.

Reply:
@merelcht Maybe subclassing is the better approach then--allow surfacing the warning, as well as using the semantically-correct one? It seems to also be the approach taken by libraries like SymPy (first link I found while searching was https://docs.sympy.org/latest/modules/utilities/exceptions.html#:~:text=The%20SymPyDeprecationWarning%20class%20is%20a,a%20link%20to%20the%20documentation.)

Reply:
Subclassing is probably the nicest approach and something we can also use in the future.

Reply:
> Yes, subclassing is probably the nicest approach and something we can also use in the future.

Will be done in a follow-up PR, more consistently across Kedro.